security/pkg/k8s/controller/monitoring.go

```diff
@@ -15,78 +15,78 @@
 package controller
 
 import (
-	"github.com/prometheus/client_golang/prometheus"
+	"istio.io/istio/security/pkg/monitoring"
 )
 
 const (
 	errorlabel = "error"
 )
 
 var (
-	serviceAccountCreationCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "svc_acc_created_cert_count",
-		Help:      "The number of certificates created due to service account creation.",
-	}, []string{})
+	errorTag     = monitoring.MustCreateTag(errorlabel)
+	subsystemTag = monitoring.MustCreateTag("subsystem")
 
-	serviceAccountDeletionCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "svc_acc_deleted_cert_count",
-		Help:      "The number of certificates deleted due to service account deletion.",
-	}, []string{})
+	serviceAccountCreationCounts = monitoring.NewSum(
+		"svc_acc_created_cert_count",
+		"The number of certificates created due to service account creation.",
+		subsystemTag,
+	)
 
-	secretDeletionCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "secret_deleted_cert_count",
-		Help:      "The number of certificates recreated due to secret deletion (service account still exists).",
-	}, []string{})
+	serviceAccountDeletionCounts = monitoring.NewSum(
+		"svc_acc_deleted_cert_count",
+		"The number of certificates deleted due to service account deletion.",
+		subsystemTag,
+	)
 
-	csrErrorCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "csr_err_count",
-		Help:      "The number of errors occurred when creating the CSR.",
-	}, []string{})
+	secretDeletionCounts = monitoring.NewSum(
+		"secret_deleted_cert_count",
+		"The number of certificates recreated due to secret deletion (service account still exists).",
+		subsystemTag,
+	)
 
-	certSignErrorCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "csr_sign_err_count",
-		Help:      "The number of errors occurred when signing the CSR.",
-	}, []string{errorlabel})
+	csrErrorCounts = monitoring.NewSum(
+		"csr_err_count",
+		"The number of errors occurred when creating the CSR.",
+		subsystemTag,
+	)
+
+	certSignErrorCounts = monitoring.NewSum(
+		"csr_sign_err_count",
+		"The number of errors occurred when signing the CSR.",
+		subsystemTag,
```

Comment:
I'm not super familiar with OpenCensus or Prometheus, but is it right? We created a new sum type metric with `subsystemTag`, then at https://github.com/istio/istio/pull/15223/files#diff-20057f9306a9c65d483295b9aa6d0ea3R91 we created a new metric with a new tag? I think `With()` in Prometheus and `With()` in our OpenCensus Metric interface are different.

Reply:
I'm also not too familiar with Prometheus or OpenCensus, but check the way OC metrics were introduced to Pilot [in this file](https://github.com/douglas-reid/istio/blob/f17fc3c0d6f584da956c4342c696a4d570852f53/pilot/pkg/proxy/envoy/discovery.go). Looks like you first [define the metric with a named tag](https://github.com/douglas-reid/istio/blob/f17fc3c0d6f584da956c4342c696a4d570852f53/pilot/pkg/proxy/envoy/discovery.go#L40), and then use `.With()` to fill in the unset tag values later.

Reply:
Oh nvm, in this case I think you're right. `With()` is to create values for this tag and update the metric :)